CI: Automatically trigger framework build by tag

bisgardo opened the pull request:
To reduce the number of steps (and things to do wrong) in the release process, the workflow is updated to automatically trigger once a tag of form `build/*.*.*-*` is pushed. The workflow will then extract the version and run by itself.

A check that the version embedded in the tag actually matches the version of the crate was also included.

bisgardo committed:
CI: Automatically trigger framework build by tag

bisgardo commented:
@soerenbf This is an attempt to make the release process a little easier. Do you think it's worth adding?

bisgardo commented on .github/workflows/publish-framework.yml:
```diff
@@ -16,12 +18,37 @@ jobs:
+        version: |
+          $(cargo metadata --format-version=1 | jq -r '.packages[] | select(.name == "concordium-wallet-crypto-uniffi") | .version')
+    - name: Fail if version in tag doesn't match crate version
+      if: "steps.crate-version.version != steps.tag.outputs.version"
```
```suggestion
      if: "steps.crate-version.outputs.version != steps.tag.outputs.version"
```

soerenbf commented on README.md:
```diff
@@ -125,21 +125,21 @@ Steps for building and releasing a new version `<version>` of the package:
-   If the build fails, commit a fix to the branch and return to step 3 with `<build-version>` incremented by 1.
-5. Run `make swift-bindings` locally to regenerate the Swift bridge sources.
-6. Update `Package.swift` with the updated `url` and `checksum` of the binary framework.
+   The pushing of this tag will automatically trigger a [workflow](./.github/workflows/publish-framework.yml)
```
```suggestion
   Pushing this tag will automatically trigger a [workflow](./.github/workflows/publish-framework.yml)
```

soerenbf reviewed (approved):
Looks good 😄 . I think (as a comment to the description) that if we have build tags specifically for the purpose of doing this thing, we might as well automate. I still think there are a lot of quirky steps to follow in order to do this but if there is no way around it, I guess it is what it is 🤷

bisgardo committed:
Update README.md

Co-authored-by: Søren Bruus Zeppelin <soerenbf@gmail.com>

bisgardo committed:
Update .github/workflows/publish-framework.yml

bisgardo commented:
> Looks good 😄 . I think (as a comment to the description) that if we have build tags specifically for the purpose of doing this thing, we might as well automate. I still think there are a lot of quirky steps to follow in order to do this but if there is no way around it, I guess it is what it is 🤷

It could probably be wrapped into fewer, more customized actions. But it's probably quite a lot of work to do in a good way.

Also the "tag override" think isn't really necessary, so I can simplify it a little by removing that.

bisgardo committed:
Merge branch 'main' into ci/framework-trigger-build-by-tag

bisgardo committed:
Finish up

Removed the "tag override" - it seems unlikely that you'll ever want to run the workflow from a different revision than the source we build from.
